numpy/lib/function_base.py

```diff
@@ -241,6 +241,39 @@ def iterable(y):
     return True
 
 
+def _hist_bin_ise(x):
+    """
+    Histogram bin estimator based on minimizing the estimated integrated squared error (ISE).
+
+    The number of bins is chosen by minimizing the estimated ISE against the unknown true distribution.
+    The ISE is estimated using cross-validation and can be regarded as a generalization of Scott's rule.
+    https://en.wikipedia.org/wiki/Histogram
+
+    Parameters
+    ----------
+    x : array_like
+        Input data that is to be histogrammed, trimmed to range. May not
+        be empty.
+
+    Returns
+    -------
+    h : An estimate of the optimal bin width for the given data.
+    """
+
+    n = x.size
+    ptp_x = np.ptp(x)
+    if n <= 1 or ptp_x == 0:
+        return ptp_x
+
+    def jhat(x, nbins):
+        hh = ptp_x / nbins
+        return (2 - (n + 1) * (np.histogram(x, bins=nbins)[0] ** 2).sum() / (n * n)) / (n - 1) / hh
```

Comment:
`N_k = np.histogram(x, bins=nbins)[0]` followed by `N_k.dot(N_k)` is likely faster than `(N_k**2).sum()`